### PR TITLE
[doc] Improve C++ Api documentation of make_zero_tail, fit_tail and fit_hermitian_tail

### PR DESCRIPTION
**Pull request by Wentzell**
* Major code refactoring for fit_tail, fit_hermitian_tail
* Clean is_gf, is_gf_v traits
* Rename trait is_block_gf_or_view -> is_block_gf
* Add trait is_block_gf_v

### CODE REVIEW

File: triqs/gfs/block/block_gf.mako.hpp [hunk 1]
```diff
@@ -41,23 +41,23 @@ namespace triqs {
     /// ---------------------------  traits ---------------------------------
 
     // Is G a block_gf, block_gf_view, block_gf_const_view
-    // is_block_gf_or_view<G> is true iif G is a block_gf or block2_gf
-    // is_block_gf_or_view<G,1> is true iff G is a block_gf
-    // is_block_gf_or_view<G,2> is true iff G is a block2_gf
+    // is_block_gf<G> is true iif G is a block_gf or block2_gf
+    // is_block_gf<G,1> is true iff G is a block_gf
+    // is_block_gf<G,2> is true iff G is a block2_gf
     //
```
Comment (parcollet) [inline triqs/gfs/block/block_gf.mako.hpp:47]: To be discussed : is_gf, is_gf_regular, is_gf_view
Container vs concept...

File: triqs/gfs/functions/functions2.hpp [hunk 2]
```diff
@@ -27,134 +27,145 @@ namespace triqs::gfs {
...
+   *
+   * @param g The Green function object to fit the tail for 
+   * @param known_moments The object containing the known high-frequency moments
+   *
+   * @return A pair of the tail object and the fitting error
+   * @example triqs/gfs/fit_tail.cpp
+   */
+  template <int N = 0, typename G, typename A = typename G::data_t::const_view_type>
+  std::pair<typename A::regular_type, double> fit_tail(G const &g, A const &known_moments = {}) REQUIRES(is_gf_v<G>) {
+    if constexpr (std::is_base_of<tag::composite, typename G::mesh_t>::value) { // product mesh
```
Comment (parcollet) [inline triqs/gfs/functions/functions2.hpp]: composite --> there is a trait 
is_product_v<MESH>

File: triqs/gfs/functions/functions2.hpp [hunk 3]
```diff
@@ -216,7 +227,7 @@ namespace triqs::gfs {
   template <typename G> std::enable_if_t<is_gf<G>::value, bool> is_gf_real(G const &g, double tolerance = 1.e-13) {
     return max_element(abs(imag(g.data()))) <= tolerance;
   }
-  template <typename G> std::enable_if_t<is_block_gf_or_view<G>::value, bool> is_gf_real(G const &g, double tolerance = 1.e-13) {
+  template <typename G> std::enable_if_t<is_block_gf<G>::value, bool> is_gf_real(G const &g, double tolerance = 1.e-13) {
```
Comment (parcollet) [inline triqs/gfs/functions/functions2.hpp]: REQUIRES ?

File: triqs/gfs/functions/functions2.hpp [hunk 4]
```diff
@@ -229,7 +240,7 @@ namespace triqs::gfs {
   template <typename G> std::enable_if_t<is_gf<G>::value, typename G::regular_type::real_t> real(G const &g) {
     return {g.mesh(), real(g.data()), g.indices()};
   }
-  template <typename G> std::enable_if_t<is_block_gf_or_view<G>::value, typename G::regular_type::real_t> real(G const &g) {
+  template <typename G> std::enable_if_t<is_block_gf<G>::value, typename G::regular_type::real_t> real(G const &g) {
```
Comment (parcollet) [inline triqs/gfs/functions/functions2.hpp]: REQUIRES ?

**Review by parcollet — COMMENTED** [review]
Fine (even though the diff is sometimes hard to follow, but it seems ok), 
 - a few points to discuss, in particular  naming of the traits (for the concept, the containers
 - a few minor simplifications.

File: triqs/gfs/functions/functions2.hpp (hunk 2 above)
Comment (Wentzell) [reply triqs/gfs/functions/functions2.hpp]: This was only introduced on the mesh_refactor branch. I would refrain from backporting this.
Do you agree?

Comment (parcollet) [reply triqs/gfs/functions/functions2.hpp]: yes

File: triqs/gfs/block/block_gf.mako.hpp (hunk 1 above)
Comment (Wentzell) [reply triqs/gfs/block/block_gf.mako.hpp:47]: I have added a new template meta function `is_instantiation_of<TMPL, TYPE>`
`is_gf_regular` and `is_gf_view` can now be written as easily as `is_instantiation_of<gf, G>`.
I have kept the behavior of the `is_gf_v` trait the same for now, i.e. not incorporating `gf_expr` and allowing for a check of the underlying `variable_t`.
We can refine the concept traits further for the 2.2 release. Can you confirm the PR?
